Refactor remove images lock

DennisDenuto opened the pull request:
The goal of this PR is to consolidate the 'localization' and 'image lock re-writing' into ImageRefs and removing the bundle 'imagelock' struct.

- ImageRefs has an ImagesLock (not the same as the above which is a now removed bundle ImagesLock) and a LocationConfig 
- UI logic when pulling a bundle has also been moved. This reduces the need of using a NoOp logger. Instead the top of the recursion stack (root bundle) is responsible for printing out logs about when the imageslock file is re-written.

DennisDenuto committed:
refactor: Remove bundle.ImagesLock

- bundle.ImagesLock is 'replaced' with existing bundle.ImagesRef.
-- ImagesRef has the responsibility of relocating all the image refs for
a particular bundle. It takes as an input an lockconfig.ImagesLock
(local file) and a ImageLocationsConfig (remote OCI). It outputs a
potentially re-written lockconfig.ImagesLock.
- rename files -> images_lock -> images_refs
- remove unused functions
- minor inlining

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
refactor: Move ui logic for pulling a root bundle

- Removes the need for using a Noop logger for sub bundles. Instead the
top of the recursion stack (root bundle) will be responsible for
printing out ImageLock log messages.

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/bundle/bundle.go:
```diff
@@ -135,32 +145,36 @@ func (o *Bundle) pull(baseOutputPath string, ui goui.UI, pullNestedBundles bool,
 	}
 
-	lock := NewImagesLock(imagesLock, o.imgRetriever, o.Repo(), LocationsConfig{
+	bundleImageRefs, err := NewImageRefs(imagesLock, LocationsConfig{
```
NewImageRefs -> NewImageRefsFromImagesLock since you can imagine plain NewImageRefs constructor wouldnt have to rely on images lock

DennisDenuto replied:
I don't understand the proposed function name change here. Today a bundle.ImageRefs is a bunch of image references from an images lock *and* a Location OCI. 

ImageRefs are *also* used to keep track of image references that have been 'processed'. But the original source of these image references came from an images lock *and* a Location OCI.

Also this is also 'relying' on a LocationsConfig which the proposed naming change doesn't capture...

cppforlife replied:
imageRefs is a just a collection of refs. one could imagine to be able to instantiate that plain collection via default constructor aka NewImageRefs() (i see that you called it NewEmptyImageRefs elsewhere, but we dont typically call other constructors NewEmpty* since that's typically the default state e.g. NewEmptyArray() vs NewArray()).

in this case we are instantiating such collection of ImageRefs _plus_ populating it with content from ImagesLock (primary source). technically, we are also using locations data to augment ImagesLock images but it's not the primary source of data, hence just the name NewImageRefsFromImagesLock().

DennisDenuto replied:
ok thanks for the explanation. renamed the function NewImageRefsFromImagesLock

cppforlife commented on pkg/imgpkg/bundle/bundle_images_lock.go:
```diff
@@ -53,7 +53,7 @@ func (o *Bundle) buildAllImagesLock(throttleReq *util.Throttle, processedImgs *p
 	}
 
-	processedImageRefs := NewImageRefs()
+	processedImageRefs := NewEmptyImageRefs()
```
NewEmptyImageRefs -> NewImageRefs()

DennisDenuto replied:
hmm `NewImageRefs` function name is already taken and used to create an ImageRefs with its dependencies.

cppforlife replied:
(see above naming comment why these names should be flipped.)

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+func (i ImageRef) DeepCopy() ImageRef {
+	return ImageRef{
+		ImageRef: i.ImageRef.DeepCopy(),
+		IsBundle: i.IsBundle,
```
pointer is being copied. need to deref and make a new pointer.

DennisDenuto replied:
ok done.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+		IsBundle: i.IsBundle,
+	}
+}
+func (i *ImageRef) updateRepo(relativeToRepo string) *ImageRef {
```
newline before

DennisDenuto replied:
i'm curious... are these new line comments 'nits' or do you see these as required changes for this PR?

cppforlife replied:
it's a nit since it's a small stylistic suggestion. whether it's required is a tough question. in this case it's very atypical to have bunch of functions bunched up together without a newline separator (i have not seen that done in many go code bases) and given our existing code base style i would want these to get fixed unless there is a really good reason not to. imho its good to do a pass to fix things like these before putting PR in the review state.

DennisDenuto replied:
I did take a pass through and did intentionally group the functions for the structs together. I can adjust to different code styles. I think if this is a coding style for the existing code base (which I'm learning about now) I'll see If I can incorporate that into our linter. (as part of a separate PR)

DennisDenuto replied:
i'll also add the new lines

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	i.Image = i.replaceImageRepo(relativeToRepo)
+	return i
+}
+func (i *ImageRef) replaceImageRepo(relativeToRepo string) string {
```
newline before

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	imageLocationsConfig *ImageLocationsConfig
+	originalImagesLock   lockconfig.ImagesLock
+
+	*sync.Mutex
```
mutex shouldnt be public.

DennisDenuto replied:
ok made it private

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	imageLocationsConfig *ImageLocationsConfig
+	originalImagesLock   lockconfig.ImagesLock
+
+	*sync.Mutex
```
it seems that this lock is meant for refs, so lets call it refsLock. read uses of i.refs should be using lock as well.

DennisDenuto replied:
ok done.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+func NewImageRefs(imagesLock lockconfig.ImagesLock, imagesLockLocationConfig ImagesLockLocationsConfig) (ImageRefs, error) {
```
naming is off for imagesLockLocationConfig. ImagesLock is different thing from Locations.

DennisDenuto replied:
ok i renamed the struct.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+		_, err = imgRetriever.Digest(image)
+		if err != nil {
+			if terr, ok := err.(*transport.Error); ok {
+				if _, ok := imageNotFoundStatusCode[terr.StatusCode]; ok {
```
we should probably pull this check out as a package level method:

_, ok := imageNotFoundStatusCode[terr.StatusCode]; ok

DennisDenuto replied:
ok done.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	for _, ref := range i.refs {
+		image, err := name.ParseReference(ref.updateRepo(relativeToRepo).Image)
```
it seems that updateRepo is mutating image ref. are we meaning to mutate some refs and then return false when finding one that isnt found? if our goal is to just get an updated location, then we should just use replaceImageRepo.

DennisDenuto replied:
The ref in the for loop is copied by value. `ref.updateRepo` does 'mutate' a locally copied ref value, but `i.refs` does not contain that change. fwiw updateRepo also only mutates the `.Image` string field on an ImageRef and doesn't update any pointers or maps etc... 

if we return false because an image is found to not be relocated... then `i.refs` is not mutated at all.

cppforlife replied:
i see. given this explanation whats the reasoning behind having updateRepo at all vs just doing something like `imgRef.replaceImageRepo(relativeToRepo)` like you were doing elsewhere?

DennisDenuto replied:
`updateRepo` calls `imgRef.replaceImageRepo(relativeToRepo)`.  `replaceImageRepo` doesn't mutate anything, it is similar to a `strings.Replace`. `updateRepo` updates / mutates the `.Image` repo portion on the imgref object.

(as per our below convo) I'm going to move `imgRef.replaceImageRepo` to instead be a package level function that accepts 2 arguments.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+
+	return true, nil
+}
+func (i *ImageRefs) AddImagesRef(refs ...ImageRef) {
```
newline

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+		}
+
+		if !found {
+			i.refs = append(i.refs, ref)
```
ref.DeepCopy()

DennisDenuto replied:
ok done!

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+		}
+	}
+}
+func (i *ImageRefs) Find(ref string) (ImageRef, bool) {
```
newline

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+func (i *ImageRefs) Find(ref string) (ImageRef, bool) {
+	for _, imageRef := range i.refs {
+		if imageRef.Image == ref {
+			return imageRef, true
```
should we be doing DeepCopy() here?

DennisDenuto replied:
I don't see why.

cppforlife replied:
in general its typically good to return copies of objects to avoid accidental mutation down the line. in this case each ImageRef contains some mutable data which we wouldnt want any caller to modify (e.g. AddLocations on lockconfig.ImageRef for example) and affect internal state of ImageRefs struct.

DennisDenuto replied:
i'll make the change.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+
+	return ImageRef{}, false
+}
+func (i *ImageRefs) MarkAsBundle(image string, isBundle bool) {
```
newline

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+		}
+	}
+}
+func (i ImageRefs) ImagesLock() lockconfig.ImagesLock {
```
newline

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+}
+func (i ImageRefs) ImagesLock() lockconfig.ImagesLock {
+	imgLock := lockconfig.NewEmptyImagesLock()
+	for _, originalImg := range i.originalImagesLock.Images {
```
lets turn i.originalImagesLock into a pointer and panic with internal inconsistency message if it's nil. we always working against a real ImagesLock

DennisDenuto replied:
I understand what is being asked. I don't understand the 'why'

cppforlife replied:
> we always working against a real ImagesLock

to expand on this: it seems that the intent for this class is to be able to generate ImagesLock based on _existing images lock_ + image refs. using this method against an ImageRefs instance that was never initialized with ImagesLock should be a programmer's mistake because we know that we must always be basing new images lock on existing images lock -- double-checking system invariants is generally a good pattern (unless its in the middle of the perf sensitive code). currently if this method is called on NewEmptyImagesRefs, it will just return empty ImagesLock. it's probably better to blow up to catch that or any other possibility instead of generating erroneous ImagesLock. this change ultimately firms up the contract of this class.

DennisDenuto replied:
Thanks for the explanation. good point i'll make the change.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	return imgLock
+}
+func (i ImageRefs) ImageRefs() []ImageRef {
+	return i.refs
```
i think we should be deep copying everything coming out of this class.

DennisDenuto replied:
why?

cppforlife replied:
(same as above deep copy comment)

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+}
+func (i *ImageRefs) syncImageRefsWithLocationConfig(imagesLockLocationConfig ImagesLockLocationsConfig) error {
+	locationsConfig, err := imagesLockLocationConfig.Config()
+	if _, ok := err.(*LocationsNotFound); ok {
+		return nil
+	}
```
would be good to move this if into below if err != nil. would read better imho.

DennisDenuto replied:
I prefer the flatter structure of these 2 different error scenarios. it has helped me differentiate the difference in return values based on these different classes of errors. 1 is a genuine error (something went wrong), the other is an error that represents something *not* going wrong, but that a bundle doesn't have a Location OCI.

cppforlife replied:
ok sounds good.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+func (i *ImageRefs) LocalizeToBundle(relativeToRepo string) {
+	for _, imgRef := range i.refs {
+		imgRef.AddLocation(imgRef.replaceImageRepo(relativeToRepo))
+		i.AddImagesRef(imgRef)
```
`i.AddImagesRef(imgRef)` feels very weird here. it reads like we are double adding refs. since we have access to index in the loop, lets do i.refs[i] = imgRef

cppforlife replied:
arguably we should just have AddLocation be mutable, so there is no need to screw around with refs

DennisDenuto replied:
Ok i'm hoping to make this feel less weird by `AddLocation` now mutates the ref in the loop.

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	return imageRefs, nil
+}
+
+func (i *ImageRefs) LocalizeToBundle(relativeToRepo string) {
```
LocalizeToBundle -> LocalizeToRepo. ImageRefs doesnt know anything bundles.

DennisDenuto replied:
ok done!

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,188 @@
+	i.Image = i.replaceImageRepo(relativeToRepo)
+	return i
+}
+func (i *ImageRef) replaceImageRepo(relativeToRepo string) string {
```
replaceImageRepo -> ImageRelativeToRepo (public). we arent really replacing anything, just calculating location. public because its used outside of ImageRef class.

DennisDenuto replied:
I disagree. I think we are replacing something. Specifically we are replacing the image's repository portion of its name with the argument provided.

I'm seeing this function being similar to `strings.Replace` where it takes an input and returns a copy of the input, replaced with something else.

It currently lives on the ImageRef struct. I think it could also live as a package function. And take 2 inputs. 1) an Image (string) and 2) relativeToRepo (string).

cppforlife replied:
i see, so the replacement is really related to image contents rather than related to ImageRef structure itself. my default instinct was that it's related to ImageRef since that's where it's at.

> It currently lives on the ImageRef struct. I think it could also live as a package function. And take 2 inputs. 1) an Image (string) and 2) relativeToRepo (string).

i would agree with you on this, either on ImageRefs or package.

DennisDenuto replied:
went with package

DennisDenuto committed:
Improve naming and improve defensive cases

- Return copies of image references
- panic if an ImageRefs does not have ref to an imageslock file. This is
an invariant and likely a programmer error.
- lock when reading imagerefs

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/bundle/images_refs.go:
```diff
@@ -0,0 +1,222 @@
+	}
+
+	for _, ref := range i.refs {
+		image, err := name.ParseReference(ref.updateRepo(relativeToRepo).Image)
```
so should this change to just using replaceImageRepo and then we remove updateRepo() method?

DennisDenuto replied:
ok done.
